src/rust/engine/ui/src/display.rs

```diff
@@ -275,6 +283,31 @@ impl EngineDisplay {
     self.flush().expect("could not flush terminal!");
   }
 
+  fn handle_stdin(&mut self) {
+    use termion::event::{Event, Key, parse_event};
+    //This buffer must have non-zero size because termion's `read` implementation for
+    //AsyncReader will return early without doing anything if it is of length 0.
+    //(See https://docs.rs/termion/1.5.4/src/termion/async.rs.html#69 )
+    let mut buf: [u8; 32] = [0; 32];
+
+    match self.async_stdin.read(&mut buf) {
+      Ok(0) => (),
+      Ok(n) => {
+        self.log(format!("Reading {} bytes: {:?}", n, buf));
+        let initial_byte: u8 = buf[0];
+        let mut iter = buf[1..].iter().map(|byte| Ok(*byte));
+        let event_or_err = parse_event(initial_byte, &mut iter);
+        match event_or_err {
+          Ok(Event::Key(Key::Ctrl('c'))) => self.log("Ctrl-C, should quit".to_string()),
```

Comment:
Looking at how other applications do it, I think this is the correct way to detect this:
For instance, https://github.com/imsnif/bandwhich/ registers a keyboard event loop [here](https://github.com/imsnif/bandwhich/blob/master/src/main.rs#L77) immediately before going into raw mode [here](https://github.com/imsnif/bandwhich/blob/master/src/main.rs#L83). Then, it reads from that event loop detecting that exact combination of modifiers [here](https://github.com/imsnif/bandwhich/blob/master/src/main.rs#L212).

Reply:
To handle exiting once we detect Ctrl-C, the best way I can think of is to create an enum RenderResult, with values {Ok(()), Err(String), Ok(Signal)} or something like that, and return that from render. Then, we can propagate that up to the Scheduler::run method’s current loop, and use the mechanism it already has to handle exits (https://github.com/blorente/pants/blob/998e4edbe31c64aeacac5142ad90433fce9a748b/src/rust/engine/src/scheduler.rs#L367)